### DISCUSSION TITLE
Nested Brackets and Ampersand usage in Tailwind UI examples

### DISCUSSION THREAD

**Discussion by aoandrew** [tailwindlabs/tailwindcss#9298] [Help] [answered]
Hi Tailwind community, can I get somebody's help in understanding the nested bracket and & usage in a few Tailwind UI examples?

I first tried asking this question on Stackoverflow [here](https://stackoverflow.com/questions/73666015/nested-brackets-and-ampersand-usage-in-tailwind-ui-examples) 

^ The above has example usage from from Tailwind UI.

Thank you.

**Comment by wongjn** [comment, marked as answer]
What you are seeing is the usage of [Tailwind's arbitrary variants feature](https://tailwindcss.com/docs/hover-focus-and-other-states#using-arbitrary-variants).

Inside the square brackets is a CSS selector (or media query, I won't be talking about that here), with the ampersand (`&`) a substitute for the element that the class is being applied to. The nested brackets are attribute selectors, like with standard CSS selectors, as you correctly inferred in your Stacked Overflow question. Underscore is used instead of spaces in a CSS selector.

To give a few examples:

```html
<div class="foo">
  <div class="[.foo_&]:text-white"></div>
</div>
```
is conceptually like:

```html
<div class="foo">
  <div class="bar"></div>
</div>
<style>
/**
 * .foo .bar
 *   ↓
 * .foo &
 *   ↓
 * .foo_&
 */
.foo .bar {
  color: white;
}
</style>
```
With `.bar` now `&`, since `.bar` is the element we are applying the class/style to.

---

Another example:

```html
<div class="foo"></div><div class="[.foo+&]:text-white"></div>
```
is conceptually like:

```html
<div class="foo"></div><div class="bar"></div>
<style>
/**
 * .foo + .bar
 *   ↓
 * .foo + &
 *   ↓
 * .foo+&
 */
.foo + .bar {
  color: white;
}
</style>
```
The `+` grammar in CSS selectors does not strictly need spaces around it, so we can remove them for a more concise arbitrary variant syntax.

---

```html
<div data-foo class="[&[data-foo]]:text-white"></div>
```
is conceptually like:

```html
<div data-foo class="bar"></div>
<style>
/**
 * .bar[data-foo]
 *   ↓
 *    &[data-foo]
 */
.bar[data-foo] {
  color: white;
}
</style>
```
Hopefully this example makes the nested brackets clear.

**Reply by aoandrew** [reply]
@wongjn really great exposition - thank you.
------------

After some time looking through your explanation - I was thinking to myself - why did Tailwind set this up? 

And then I realized, wow, it's a way to "backdoor" code-like specificity directly into the class (1/removes the need to set up various child / parent, etc.  relationship in a stylesheet and 2/can potentially remove the need to add classes to specific elements within your code.)

**Reply by aoandrew** [reply]
Also @wongjn if you want to put your answer in stackoverflow - I'm happy to give you credit.   Cheers.

Again, appreciate the answer!  If you don't see this in a couple days, I'll just answer it so others can benefit, and tag your github handle somewhere in the answer.
